From 0f05f4860118ffc903819ca7d354d4d63b5123aa Mon Sep 17 00:00:00 2001
From: Gokul Sriram Palanisamy <quic_gokulsri@quicinc.com>
Date: Tue, 21 May 2024 11:03:24 +0530
Subject: [PATCH] bus: mhi: host: enable netdev driver

address Linux 6.1 compatibility fixes on downstream netdev driver.
Also enable RMNET CTL channel.

Change-Id: Ic9f22ff8dabcee6333e5f83acfc394837d66a4b5
Signed-off-by: Gokul Sriram Palanisamy <quic_gokulsri@quicinc.com>
---

diff --git a/drivers/bus/mhi/clients/netdev.c b/drivers/bus/mhi/clients/netdev.c
index 428d53f..a7bdfc2 100644
--- a/drivers/bus/mhi/clients/netdev.c
+++ b/drivers/bus/mhi/clients/netdev.c
@@ -9,7 +9,6 @@
 #include <linux/if_arp.h>
 #include <linux/dma-mapping.h>
 #include <linux/debugfs.h>
-#include <linux/ipc_logging.h>
 #include <linux/device.h>
 #include <linux/errno.h>
 #include <linux/of_device.h>
@@ -86,6 +85,12 @@
 		      enum dma_data_direction dir);
 };
 
+struct mhi_device_info {
+	const char *ifname;
+	u32 mru;
+	bool chain;
+};
+
 static struct mhi_driver mhi_netdev_driver;
 static void mhi_netdev_create_debugfs(struct mhi_netdev *mhi_netdev);
 static u32 napi_poll_weight = NAPI_POLL_WEIGHT;
@@ -204,7 +209,7 @@
 	struct mhi_netbuf *netbuf;
 	struct mhi_buf *mhi_buf;
 	struct mhi_netbuf **netbuf_pool = mhi_netdev->netbuf_pool;
-	int nr_tre = mhi_get_no_free_descriptors(mhi_dev, DMA_FROM_DEVICE);
+	int nr_tre = mhi_get_free_desc_count(mhi_dev, DMA_FROM_DEVICE);
 	int i, peak, cur_index, ret;
 	const int pool_size = mhi_netdev->pool_size - 1, max_peak = 4;
 
@@ -546,9 +551,7 @@
 	if (mhi_netdev->alias < 0)
 		mhi_netdev->alias = 0;
 
-	ret = of_property_read_string(of_node, "mhi,interface-name",
-				      &mhi_netdev->interface_name);
-	if (ret)
+	if (!mhi_netdev->interface_name)
 		mhi_netdev->interface_name = mhi_netdev_driver.driver.name;
 
 	if (!strcmp(mhi_netdev->interface_name, MHI_RMNET_IF_NAME))
@@ -585,7 +588,7 @@
 		goto napi_alloc_fail;
 	}
 
-	netif_napi_add(mhi_netdev->ndev, mhi_netdev->napi,
+	netif_napi_add_weight(mhi_netdev->ndev, mhi_netdev->napi,
 		       mhi_netdev_poll, napi_poll_weight);
 	ret = register_netdev(mhi_netdev->ndev);
 	if (ret) {
@@ -911,11 +914,11 @@
 	return single_open(file, fwainfo_proc_show, NULL);
 }
 
-static const struct file_operations fwainfo_proc_fops = {
-	.open		= fwainfo_proc_open,
-	.read		= seq_read,
-	.llseek		= seq_lseek,
-	.release	= single_release,
+static const struct proc_ops fwainfo_proc_fops = {
+	.proc_open		= fwainfo_proc_open,
+	.proc_read		= seq_read,
+	.proc_lseek		= seq_lseek,
+	.proc_release		= single_release,
 };
 
 static int mhi_netdev_probe(struct mhi_device *mhi_dev,
@@ -927,11 +930,9 @@
 	struct device *dbg_dev = &mhi_dev->dev;
 	int nr_tre;
 	struct device_node *phandle;
-	bool no_chain;
+	bool chain;
 	static int fwa_proc_created;
-
-	if (!of_node)
-		return -ENODEV;
+	const struct mhi_device_info *info = (struct mhi_device_info *)id->driver_data;
 
 	mhi_netdev = devm_kzalloc(&mhi_dev->dev, sizeof(*mhi_netdev),
 				  GFP_KERNEL);
@@ -941,9 +942,9 @@
 	mhi_netdev->mhi_dev = mhi_dev;
 	dev_set_drvdata(&mhi_dev->dev, mhi_netdev);
 
-	ret = of_property_read_u32(of_node, "mhi,mru", &mhi_netdev->mru);
-	if (ret)
-		return -ENODEV;
+	mhi_netdev->interface_name = info->ifname;
+	mhi_netdev->mru = info->mru;
+	chain = info->chain;
 
 	/* MRU must be multiplication of page size */
 	mhi_netdev->order = __ilog2_u32(mhi_netdev->mru / PAGE_SIZE);
@@ -973,9 +974,7 @@
 		mhi_netdev_clone_dev(mhi_netdev, p_netdev);
 		put_device(dev);
 	} else {
-		no_chain = of_property_read_bool(of_node,
-						 "mhi,disable-chain-skb");
-		if (!no_chain) {
+		if (chain) {
 			mhi_netdev->chain = devm_kzalloc(&mhi_dev->dev,
 						sizeof(*mhi_netdev->chain),
 						GFP_KERNEL);
@@ -998,7 +997,7 @@
 	}
 
 	/* setup pool size ~2x ring length*/
-	nr_tre = mhi_get_no_free_descriptors(mhi_dev, DMA_FROM_DEVICE);
+	nr_tre = mhi_get_free_desc_count(mhi_dev, DMA_FROM_DEVICE);
 	mhi_netdev->pool_size = 1 << __ilog2_u32(nr_tre);
 	if (nr_tre > mhi_netdev->pool_size)
 		mhi_netdev->pool_size <<= 1;
@@ -1043,11 +1042,23 @@
 	return ret;
 }
 
+static const struct mhi_device_info mhi_hwip0 = {
+	.ifname = "rmnet_mhi",
+	.mru = 0x4000,
+	.chain = true
+};
+
+static const struct mhi_device_info mhi_swip0 = {
+	.ifname = "rmnet_mhi_sw",
+	.mru = 0x4000,
+	.chain = false
+};
+
 static const struct mhi_device_id mhi_netdev_match_table[] = {
-	{ .chan = "IP_HW0" },
+	{ .chan = "IP_HW0", .driver_data = (kernel_ulong_t)&mhi_hwip0 },
 	{ .chan = "IP_HW_ADPL" },
 	{ .chan = "IP_HW0_RSC" },
-	{ .chan = "IP_SW0" },
+	{ .chan = "IP_SW0", .driver_data = (kernel_ulong_t)&mhi_swip0 },
 	{},
 };
 
diff --git a/drivers/bus/mhi/host/init.c b/drivers/bus/mhi/host/init.c
index 9671e38..0b9ba7e 100644
--- a/drivers/bus/mhi/host/init.c
+++ b/drivers/bus/mhi/host/init.c
@@ -1324,6 +1324,7 @@
 	}
 
 	mhi_dev->mhi_cntrl = mhi_cntrl;
+	mhi_dev->mtu = MHI_MAX_MTU;
 	mhi_dev->dev_wake = 0;
 
 	return mhi_dev;
diff --git a/drivers/bus/mhi/host/pci_generic.c b/drivers/bus/mhi/host/pci_generic.c
index caa4ce2..9975347 100644
--- a/drivers/bus/mhi/host/pci_generic.c
+++ b/drivers/bus/mhi/host/pci_generic.c
@@ -223,6 +223,20 @@
 		.channel = ch_num,		\
 	}
 
+#define MHI_EVENT_CONFIG_HW_DATA_CLI_MANAGED(ev_ring, el_count, ch_num) \
+	{					\
+		.num_elements = el_count,	\
+		.irq_moderation_ms = 1,		\
+		.irq = (ev_ring) + 1,		\
+		.priority = 1,			\
+		.mode = MHI_DB_BRST_DISABLE,	\
+		.data_type = MHI_ER_DATA,	\
+		.hardware_event = true,		\
+		.client_managed = true,	\
+		.offload_channel = false,	\
+		.channel = ch_num,		\
+	}
+
 static const struct mhi_channel_config modem_qcom_v1_mhi_channels[] = {
 	MHI_CHANNEL_CONFIG_UL(4, "DIAG", 16, 1),
 	MHI_CHANNEL_CONFIG_DL(5, "DIAG", 16, 1),
@@ -236,6 +250,8 @@
 	MHI_CHANNEL_CONFIG_DL_FP(35, "FIREHOSE", 32, 0),
 	MHI_CHANNEL_CONFIG_HW_UL(100, "IP_HW0", 128, 2),
 	MHI_CHANNEL_CONFIG_HW_DL(101, "IP_HW0", 128, 3),
+	MHI_CHANNEL_CONFIG_UL(105, "RMNET_CTL", 128, 4),
+	MHI_CHANNEL_CONFIG_DL(106, "RMNET_CTL", 128, 5),
 };
 
 static struct mhi_event_config modem_qcom_v1_mhi_events[] = {
@@ -245,7 +261,9 @@
 	MHI_EVENT_CONFIG_DATA(1, 128),
 	/* Hardware channels request dedicated hardware event rings */
 	MHI_EVENT_CONFIG_HW_DATA(2, 1024, 100),
-	MHI_EVENT_CONFIG_HW_DATA(3, 2048, 101)
+	MHI_EVENT_CONFIG_HW_DATA_CLI_MANAGED(3, 2048, 101),
+	MHI_EVENT_CONFIG_HW_DATA(4, 1024, 105),
+	MHI_EVENT_CONFIG_HW_DATA(5, 1024, 106),
 };
 
 static const struct mhi_controller_config modem_qcom_v1_mhiv_config = {
diff --git a/include/linux/mhi.h b/include/linux/mhi.h
index 7a74403..3d21cfe1 100644
--- a/include/linux/mhi.h
+++ b/include/linux/mhi.h
@@ -490,6 +490,7 @@
 	int ul_chan_id;
 	int dl_chan_id;
 	u32 dev_wake;
+	size_t mtu;
 };
 
 /**
